### ISSUE TITLE
Pluggable permissions

### ISSUE THREAD

**Issue by spion-h4** [hfour/h4bff#43]
Make a permission class that is easy to modify from the outside.

```typescript

class DocumentPermissions extends AppSingleton {
  read = new Permission<Document>();
  write = new Permission<Document>();
}
getSingleton(DocumentPermissions).write
  .add({name: string, priority: level, check: (doc: Document, ctx: ServiceContext) => PermisisonResult})

type PermissionResult = 'allow' | 'deny' | 'pass'
```

Algorithm: Sort by priority, the first check function that returns “Allow” or “Deny” instead of “Pass” wins.

Desireables:

easy way to query all active rules in all active permissions

easy recording of which rules were consulted and what their results were in the audit log

**Comment by spion-h4** [comment]
Building a simple role based system on top of this

```typescript
function makeRolePermission(actionName: string) {
  return { 
    name: 'Role based check', 
    priority: 10, 
    check: async (_, ctx) =>  { 
      let uid = this.getService(UserId).id;
      let roleCheck = await ctx.getService(Roles).checkRole(actionName, uid);
      return roleCheck ? 'allow' : 'deny'
    }
  } 
}
```

Considerations:
* It seems to be flexible enough for this
* Do we need the check function to be able to return additional description (for audit log) ?

**Comment by spion-h4** [comment]
Building simple permissions like cancan and pundit:

```
permission.addRule({
  name: 'Superadmin can access everything'
  priority: 100,
  check: (_, ctx) => ctx.getService(UserAdminInfo).isSuperAdmin ? 'allow' : 'pass'
}
```

Pass means the rule will defer to other rules if the user is not super admin.

Considerations
* Do we want to add the superadmin rule to everything?
* Do we use inherits a lot to avoid adding the same rule to everything?
* Will the permissions be hard to follow with inherits? (a system to query all relevant rules would be helpful)

**Comment by spion-h4** [comment]
To be able to keep track of all permissions, you could have a singleton:

```typescript
app.getSingleton(Permissions).create(name: string)
```

You could also plug into this now

```typescript
app.getSingleton(Permissions).onPermissionQueried((result, ctx) => {
  ctx.getService(AuditLogBuilder).record({permissionCheckResult: result.checkedRules})
})
```

**Comment by spion-h4** [comment]
More rule analysis: we can protect the `check` function from being directly assigned, instead create a checking object from a function with allowed return results

```typescript
check: checkFn(['allow', 'deny'], fn)
```

That way you can detect anomalies like a high priority rule is "cutting off" an entire tree of low priority ones by never returning 'pass'

**Comment by spion-h4** [comment]
Rule ordering alternative: Instead of having numbers, create subrules as "slots" in other rules and export the slot only. Then the order is strictly determined.

```typescript
export class ConfigurablePermissions extends AppSingleton {
  read = permission.create()
}

// not exported
class RealPermissions extends AppSingleton {
  private exported = app.getSingleton(ConfigurablePermissions);
  read = permission.create().add(allowAdmins, denyDeactivatedAccounts, exported.read, defaultDeny);
  write = permission.create().add(allowAdmins, denyDeactivatedAccounts, exported.write, defaultDeny);
}

// exported
export class ConfigurablePermissions extends AppSingleton {
  public read = permission.create();
  public write = permission.create();
}
```

**Comment by spion-h4** [comment]
Here is how a report tree might look like for a single permission

### Action: document.read

```
"User is super admin" -> Allow
| (pass)
"Request coming from other app, not user" -> Allow
| (pass)
"Account has been deactivated" -> Deny
| (pass)
"User is member of library" -> Allow
```

**Comment by wh4everest** [comment]
Consideration about graying out buttons on the frontend...

Currently, our frontends are littered with code that performs the same checks that the backend does, in order to disable some input and prevent a request from happening.

For example:

```typescript
if (!roles.user.includes(['editor']) { makeUploadButtonGray() }`
```

Instead, it would be nice if you can ask a backend service whether I can perform the action and receive a boolean answer.

```typescript
myService.checkPermissions.then(perms => { if (!perms.read) makeUploadButtonGray() }`
```

Behind the scenes, the service would calculate the outcome of `permission.create(ctx)` for each method and return me the object full of boolean results.

Of course, this won't work if the answer depends on the params ("can I delete this particular set of documents?") but that's fine, since you'd get 403 as a response -- rarely enough not to annoy the users.

**Comment by spion** [comment]
The role system can provide (in the rule) more details by returning

```
allowDescription: Promise<string>
denyDescription: Promise<string>
```

The results would probably look like
Allow for roles: X, Y, Z
Deny for roles: A, B


```

**Comment by spion** [comment]
How to get a typesafe permission installation for RPC methods

class MyService {
...
}

setPermissions(MyService, {
  method1: permission1
  method2: permission2
});

Advantages to decorators:
* typesafe params

Advantages to checking in method body
* don't have to invoke the method to find permissions
* can automatically build an asking service "svcName.method.askPermission`

Disadvantages
* permissions a bit far away from method

**Comment by spion-h4** [comment]
Simplified "slot" design:

```
function basePermission(additionalRules) {
  return [startRules].concat(additionRules).concat(defaultDeny);
}

function check(ruleList) {
  return x;
}

function describe(ruleList) {
  return x;
}
```
